code/modules/events/crystal_event.dm

```diff
@@ -0,0 +1,408 @@
+	var/list/sm_crystal = list()
+	for(var/obj/machinery/power/supermatter_crystal/temp in GLOB.machines)
+		if(QDELETED(temp))
+			WARNING("No engine found.")
+			kill()
+		if(istype(temp, /obj/machinery/power/supermatter_crystal/shard))
+			message_admins("found a shard, skipping it")
+			continue
+		sm_crystal += temp
+	if(sm_crystal == null)
+		WARNING("No engine found.")
+		kill()
+	var/obj/machinery/power/supermatter_crystal/crystal = pick(sm_crystal)
+	crystal.destabilize()
+
+	priority_announce("WARNING - Numerous energy fluctuations have been detected from your Supermatter; we estimate a [wave_name] of crystalline creatures \
+						coming from \[REDACTED]; there will be [portal_numbers] portals spread around the station that you must close. Harvest a \[REDACTED] \
+						anomaly from a portal, place it inside a crystal stabilizer, and inject it into your Supermatter to stop a ZK-Lambda-Class Cosmic Fragmentation Scenario from occurring.", "Alert", 'sound/misc/notice1.ogg')
+
+	sleep(10 SECONDS)
+
+	for(var/i = 0, i< portal_numbers, i++)
+		spawn_portal(wave_type)
+
+///Spawn one portal in a random location choosen from the generic_event_spawns list
+/datum/round_event/crystal_invasion/proc/spawn_portal(list/wave_type)
+	var/list/spawners = list()
+	for(var/obj/effect/landmark/event_spawn/temp in GLOB.generic_event_spawns)
+		if(QDELETED(temp))
+			continue
+		if(is_station_level(temp.loc.z))
+			spawners += temp
+
+	if(!spawners.len)
+		message_admins("No APCs on the station, aborting crystal event")
+		return MAP_ERROR
+
+	var/pick_portal = pickweight(wave_type)
+	var/obj/spawner = pick_n_take(spawners)
+	new pick_portal(spawner.loc)
+
+/*
+This section is for the destabilized SM
+*/
+
+/obj/machinery/destabilized_crystal
+	name = "destabilized crystal"
+	desc = "A strangely translucent and iridescent crystal."
+	icon = 'icons/obj/supermatter.dmi'
+	icon_state = "psy"
+	density = TRUE
+	anchored = TRUE
+	flags_1 = PREVENT_CONTENTS_EXPLOSION_1
+	light_range = 4
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF | FREEZE_PROOF
+	///If not active the crystal will not emit radiations and gases
+	var/active = TRUE
+
+/obj/machinery/destabilized_crystal/Initialize()
+	. = ..()
+
+/obj/machinery/destabilized_crystal/Destroy()
+	return..()
+
+/obj/machinery/destabilized_crystal/process()
+	if(active)
+		if(prob(45))
+			src.fire_nuclear_particle()
+			radiation_pulse(src, 250, 6)
+		var/turf/T = loc
+		var/datum/gas_mixture/env = T.return_air()
+		var/datum/gas_mixture/removed
+		var/gasefficency = 0.5
+		removed = env.remove(gasefficency * env.total_moles())
+		removed.assert_gases(/datum/gas/bz, /datum/gas/stimulum, /datum/gas/nitryl, /datum/gas/miasma)
+		removed.gases[/datum/gas/bz][MOLES] += 5.5
+		removed.gases[/datum/gas/stimulum][MOLES] += 4.5
+		removed.gases[/datum/gas/nitryl][MOLES] += 6.75
+		removed.gases[/datum/gas/miasma][MOLES] += 10.5
+		env.merge(removed)
+		air_update_turf()
```

Comment:
How slow is this in testing/how large can this get?

Reply:
I could see this making a lot of active atmos turfs

Reply:
the max this got was filling up the SM chamber with gases, it was pretty slow but i could remove some of the gases (like stim and nitryl) to remove some of the load to the atmos

Reply:
Might be a good idea in general, depending on how you want to limit access to those gasses.

Reply:
removed stim and nitryl productions